### PR TITLE
Add Support for refined endsWith #481

### PR DESCRIPTION
**Pull request by mrcmatuszak**
Basically it's similar to one from 'startsWith' Prism with fthomas comments applied.

### CODE REVIEW

**Comment by fthomas** [comment]
LGTM! If you want to go the next step of generalizing the `Prism` definitions, you could factor out a `def refinedPrism[T, P](implicit v: Validate[T, P]): Prism[T, T Refined P] = ...` and implement all four refined prisms with this function.

**Comment by mrcmatuszak** [comment]
Add rest of refined functions: MatchesRegex, Regex, Uri, Url, UUID, Xml, Xpath?

File: refined/shared/src/main/scala/monocle/refined/strings.scala [hunk 1]
```diff
@@ -8,16 +8,17 @@ object strings extends StringsInstances
...
-    Prism.partial[String, Refined[String, EndsWith[suffix.type]]] {
-      case string if v.isValid(string) => Refined.unsafeApply(string)
-    }{_.value}
-  }
+  def endsWith(suffix: String)(implicit v: Validate[String, EndsWith[suffix.type]]): Prism[String, EndsWithString[suffix.type]] =
+    refinedPrism[String, EndsWith[suffix.type]](suffix)
 
+  private def refinedPrism[T, P](t: T)(implicit v: Validate[T, P]): Prism[T, T Refined P] = {
+    Prism.partial[T, Refined[T, P]] {
+      case tt if v.isValid(t) => Refined.unsafeApply[T, P](tt)
```
Comment (fthomas) [inline refined/shared/src/main/scala/monocle/refined/strings.scala]: We need to check if `tt` is valid since this is the value we are wrapping with `Refined`. The `t` parameter here will be `prefix` and `suffix` from above and therefore `v.isValid(t)` will always be true. The `t` parameter is not needed to define `refinedPrism`. If you remove that parameter, you can also use `refinedPrism` to implement `chars.lowerCase` and `chars.upperCase` with it.

File: refined/shared/src/main/scala/monocle/refined/strings.scala [hunk 2]
```diff
@@ -8,16 +8,17 @@ object strings extends StringsInstances
 
 trait StringsInstances {
 
-  def startsWith(prefix: String): Prism[String, StartsWithString[prefix.type]] = {
-    Prism.partial[String, Refined[String, StartsWith[prefix.type]]] {
-      case string if string.startsWith(prefix) => Refined.unsafeApply(string)
-    }{_.value}
-  }
+  def startsWith(prefix: String)(implicit v: Validate[String, StartsWith[prefix.type]]): Prism[String, StartsWithString[prefix.type]] =
+    refinedPrism[String, StartsWith[prefix.type]](prefix)
```
Comment (fthomas) [inline refined/shared/src/main/scala/monocle/refined/strings.scala]: I think you can omit the type parameters to `refinedPrism` here, type inference will figure them out.

File: refined/shared/src/main/scala/monocle/refined/strings.scala (hunk 1 above)
Comment (mrcmatuszak) [inline refined/shared/src/main/scala/monocle/refined/strings.scala]: of course it should be `tt` not `t`, lame mistake.

File: refined/shared/src/main/scala/monocle/refined/strings.scala [hunk 3]
```diff
@@ -1,17 +1,24 @@
...
+
+  def endsWith(suffix: String)(implicit v: Validate[String, EndsWith[suffix.type]]): Prism[String, EndsWithString[suffix.type]] =
+    refinedPrism[String, EndsWith[suffix.type]](suffix)
 
+  private def refinedPrism[T, P](t: T)(implicit v: Validate[T, P]): Prism[T, T Refined P] = {
```
Comment (fthomas) [inline refined/shared/src/main/scala/monocle/refined/strings.scala]: Btw, if we make this public, people can use it for their own custom predicates that are not in refined.

Comment (julien-truffaut) [inline refined/shared/src/main/scala/monocle/refined/strings.scala]: Yes I think we should expose a method like that but in a more general package (I don't think it is specific to Refined String). We can do that on another PR.

**Comment by julien-truffaut** [comment]
👍 Thanks @mrcmatuszak and @fthomas for the review
